File: oop-fundamentals/7.polymorphism/cpp/polymorphism.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

// Ref - https://algomaster.io/learn/lld/polymorphism
// Ref - https://www.geeksforgeeks.org/polymorphism-in-c/
// Ref - https://en.cppreference.com/w/cpp/language/virtual

//
// =======================================================
// 1. WHAT IS POLYMORPHISM?
// =======================================================
//
// Polymorphism = "Many forms"
// It allows the same interface to be used ford different underlying forms (data
// types).
//
// Key formula:
// Polymorphism = Same Interface + Different Behaviors
//
// Benefits:
// - Code reusability (write generic code)
// - Flexibility (add new types without changing existing code)
// - Loose coupling (depend on abstractions, not concrete types)
// - Extensibility (easily add new implementations)
//
// Types in C++:
// 1. Compile-time Polymorphism (Static/Early Binding)
//    - Function Overloading
//    - Operator Overloading
//    - Templates
//
// 2. Runtime Polymorphism (Dynamic/Late Binding)
//    - Virtual Functions
//    - Function Overriding

//
// =======================================================
// 2. COMPILE-TIME POLYMORPHISM: FUNCTION OVERLOADING
// =======================================================
//
// Multiple functions with same name but different parameters
// Compiler decides which function to call based on arguments

class Calculator {
public:
  // Overload 1: Two integers
  int add(int a, int b) {
    std::cout << "📊 add(int, int)\n";
    return a + b;
  }

  // Overload 2: Three integers
  int add(int a, int b, int c) {
    std::cout << "📊 add(int, int, int)\n";
    return a + b + c;
  }

  // Overload 3: Two doubles
  double add(double a, double b) {
    std::cout << "📊 add(double, double)\n";
    return a + b;
  }

  // Overload 4: String concatenation
  std::string add(const std::string &a, const std::string &b) {
    std::cout << "📊 add(string, string)\n";
    return a + b;
  }
};

//
// =======================================================
// 3. COMPILE-TIME POLYMORPHISM: OPERATOR OVERLOADING
// =======================================================
//
// Redefine how operators work for custom types

class Point {
private:
  double x, y;

public:
  Point(double x = 0, double y = 0) : x(x), y(y) {}

  // Overload + operator
  Point operator+(const Point &other) const {
    return Point(x + other.x, y + other.y);
  }

  // Overload == operator
  bool operator==(const Point &other) const {
    return x == other.x && y == other.y;
  }

  // Overload << for output
  friend std::ostream &operator<<(std::ostream &os, const Point &p) {
    os << "(" << p.x << ", " << p.y << ")";
    return os;
  }

  double getX() const { return x; }
  double getY() const { return y; }
};

//
// =======================================================
// 4. RUNTIME POLYMORPHISM: VIRTUAL FUNCTIONS
// =======================================================
//
// Virtual functions enable dynamic dispatch
// The function called is determined at RUNTIME based on actual object type

class Shape {
protected:
  std::string color;

public:
  Shape(const std::string &c) : color(c) {}

  // Virtual function - can be overridden
  virtual double getArea() const { return 0.0; }

  virtual void draw() const {
    std::cout << "Drawing a shape with color: " << color << "\n";
  }

  // Virtual destructor - ALWAYS use for base classes!
  virtual ~Shape() { std::cout << "~Shape() destructor\n"; }
};

class Circle : public Shape {
private:
  double radius;

public:
  Circle(const std::string &c, double r) : Shape(c), radius(r) {}

  // Override base class method
  double getArea() const override { return 3.14159 * radius * radius; }

  void draw() const override {
    std::cout << "⭕ Drawing Circle | Color: " << color
              << " | Area: " << getArea() << "\n";
  }

  ~Circle() { std::cout << "~Circle() destructor\n"; }
};

class Rectangle : public Shape {
private:
  double width, height;

public:
  Rectangle(const std::string &c, double w, double h)
      : Shape(c), width(w), height(h) {}

  double getArea() const override { return width * height; }

  void draw() const override {
    std::cout << "▭ Drawing Rectangle | Color: " << color
              << " | Area: " << getArea() << "\n";
  }

  ~Rectangle() { std::cout << "~Rectangle() destructor\n"; }
};

//
// =======================================================
// 5. HOW RUNTIME POLYMORPHISM WORKS: vtable & vptr
// =======================================================
//
// When a class has virtual functions, the compiler creates:
// - vtable (virtual table): Array of function pointers for that class
// - vptr (virtual pointer): Pointer in each object to its class's vtable
//
// Runtime dispatch:
// 1. basePtr->virtualFunc()
// 2. Follow basePtr's vptr to vtable
// 3. Look up virtualFunc in vtable
// 4. Call the correct overridden version!
//
// This happens at RUNTIME, not compile time.

//
// =======================================================
// 6. PURE VIRTUAL FUNCTIONS & ABSTRACT CLASSES
// =======================================================
//
// Pure virtual function: = 0
// Abstract class: Has at least one pure virtual function
// Cannot instantiate abstract classes!

class Animal {
protected:
  std::string name;

public:
  Animal(const std::string &n) : name(n) {}

  // Pure virtual function
  virtual void makeSound() const = 0;

  // Regular virtual function
  virtual void sleep() const { std::cout << name << " is sleeping 💤\n"; }

  virtual ~Animal() = default;
};

class Dog : public Animal {
public:
  Dog(const std::string &n) : Animal(n) {}

  void makeSound() const override {
    std::cout << name << " says: Woof! Woof! 🐕\n";
  }
};

class Cat : public Animal {
public:
  Cat(const std::string &n) : Animal(n) {}

  void makeSound() const override {
    std::cout << name << " says: Meow! Meow! 🐱\n";
  }
};

//
// =======================================================
// 7. REAL-WORLD EXAMPLE: PAYMENT SYSTEM
// =======================================================
//

class PaymentMethod {
protected:
  std::string accountId;

public:
  PaymentMethod(const std::string &id) : accountId(id) {}

  virtual bool processPayment(double amount) = 0;
  virtual std::string getType() const = 0;
  virtual ~PaymentMethod() = default;
};

class CreditCard : public PaymentMethod {
private:
  std::string cardNumber;

public:
  CreditCard(const std::string &id, const std::string &card)
      : PaymentMethod(id), cardNumber(card) {}

  bool processPayment(double amount) override {
    std::cout << "💳 Processing $" << amount << " via Credit Card\n";
    std::cout << "   Card: " << cardNumber.substr(0, 4) << "****\n";
    return true;
  }

  std::string getType() const override { return "Credit Card"; }
};

class PayPal : public PaymentMethod {
private:
  std::string email;

public:
  PayPal(const std::string &id, const std::string &mail)
      : PaymentMethod(id), email(mail) {}

  bool processPayment(double amount) override {
    std::cout << "💰 Processing $" << amount << " via PayPal\n";
    std::cout << "   Email: " << email << "\n";
    return true;
  }

  std::string getType() const override { return "PayPal"; }
};

class Cryptocurrency : public PaymentMethod {
private:
  std::string walletAddress;

public:
  Cryptocurrency(const std::string &id, const std::string &wallet)
      : PaymentMethod(id), walletAddress(wallet) {}

  bool processPayment(double amount) override {
    std::cout << "₿ Processing $" << amount << " via Cryptocurrency\n";
    std::cout << "   Wallet: " << walletAddress.substr(0, 8) << "...\n";
    return true;
  }

  std::string getType() const override { return "Cryptocurrency"; }
};

//
// =======================================================
// 8. POLYMORPHIC BEHAVIOR IN ACTION
// =======================================================
//

void processTransaction(PaymentMethod *method, double amount) {
  std::cout << "\n--- Processing Transaction ---\n";
  std::cout << "Payment Type: " << method->getType() << "\n";
  method->processPayment(amount);
  std::cout << "✅ Transaction complete!\n";
}

//
// =======================================================
// 9. FUNCTION OVERRIDING VS OVERLOADING
// =======================================================
//
// OVERLOADING (Compile-time):
// - Same function name, different parameters
// - In same class
// - Resolved at compile time
//
// OVERRIDING (Runtime):
// - Same function signature in base and derived
// - Requires 'virtual' in base, 'override' in derived
// - Resolved at runtime

class Demo {
public:
  // Overloading - same name, different parameters
  void display(int x) { std::cout << "Integer: " << x << "\n"; }

  void display(double x) { std::cout << "Double: " << x << "\n"; }

  void display(const std::string &x) { std::cout << "String: " << x << "\n"; }
};

//
// =======================================================
// 10. SLICING PROBLEM & WHY USE POINTERS/REFERENCES
// =======================================================
//
// Object slicing occurs when you assign a derived object to a base object
// The derived-specific data is "sliced off"

void demonstrateSlicing() {
  std::cout << "\n--- Demonstrating Object Slicing ---\n";

  Circle circle("Red", 5);

  // ❌ Slicing: Derived object assigned to base object
  Shape shape = circle; // Circle-specific data lost!
  shape.draw();         // Calls Shape::draw(), not Circle::draw()

  std::cout << "\n--- Using Polymorphism (Pointer/Reference) ---\n";

  // ✅ Polymorphism: Base pointer to derived object
  Shape *shapePtr = &circle;
  shapePtr->draw(); // Calls Circle::draw() (polymorphic!)

  // ✅ Polymorphism: Base reference to derived object
  Shape &shapeRef = circle;
  shapeRef.draw(); // Calls Circle::draw() (polymorphic!)
}

//
// =======================================================
// 11. DEMONSTRATION
// =======================================================
//

int main() {
  std::cout << "=== Polymorphism in C++ Demo ===\n\n";

  // ---- Compile-time Polymorphism: Function Overloading ----
  std::cout << "1. Function Overloading (Compile-time):\n";
  Calculator calc;
  std::cout << "Result: " << calc.add(5, 3) << "\n";
  std::cout << "Result: " << calc.add(5, 3, 2) << "\n";
  std::cout << "Result: " << calc.add(5.5, 3.2) << "\n";
  std::cout << "Result: "
            << calc.add(std::string("Hello"), std::string(" World")) << "\n";

  // ---- Compile-time Polymorphism: Operator Overloading ----
  std::cout << "\n2. Operator Overloading (Compile-time):\n";
  Point p1(3, 4);
  Point p2(1, 2);
  Point p3 = p1 + p2;
  std::cout << "p1 = " << p1 << "\n";
  std::cout << "p2 = " << p2 << "\n";
  std::cout << "p3 = p1 + p2 = " << p3 << "\n";
  std::cout << "p1 == p2? " << (p1 == p2 ? "Yes" : "No") << "\n";

  // ---- Runtime Polymorphism: Virtual Functions ----
  std::cout << "\n3. Runtime Polymorphism (Virtual Functions):\n";
  {
    Circle circle("Blue", 5);
    Rectangle rect("Green", 10, 20);

    // Polymorphic behavior through base class pointers
    std::vector<Shape *> shapes = {&circle, &rect};

    for (Shape *shape : shapes) {
      shape->draw(); // Calls correct derived version!
    }
  }

  // ---- Pure Virtual Functions & Abstract Classes ----
  std::cout << "\n4. Abstract Classes (Pure Virtual Functions):\n";
  Dog dog("Buddy");
  Cat cat("Whiskers");

  // Animal* animal = new Animal("Generic");  // ❌ Error: cannot instantiate

  std::vector<Animal *> animals = {&dog, &cat};
  for (Animal *animal : animals) {
    animal->makeSound();
    animal->sleep();
  }

  // ---- Real-World Example: Payment System ----
  std::cout << "\n5. Real-World Example (Payment System):\n";
  CreditCard creditCard("ACC001", "1234567890123456");
  PayPal paypal("ACC002", "user@example.com");
  Cryptocurrency crypto("ACC003", "0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb");

  std::vector<PaymentMethod *> payments = {&creditCard, &paypal, &crypto};

  for (PaymentMethod *payment : payments) {
    processTransaction(payment, 100.00);
  }

  // ---- Function Overloading vs Overriding ----
  std::cout << "\n6. Function Overloading (Compile-time):\n";
  Demo demo;
  demo.display(42);
  demo.display(3.14);
  demo.display("Hello");

  // ---- Object Slicing Problem ----
  demonstrateSlicing();

  std::cout << "\n=== Demo Complete ===\n";
  return 0;
}

/*
📘 Key Insights:

1. **Two Types of Polymorphism**:
   - Compile-time: Function/Operator overloading, Templates (static dispatch)
   - Runtime: Virtual functions, Method overriding (dynamic dispatch)

2. **Virtual Functions**:
   - Use 'virtual' keyword in base class
   - Use 'override' keyword in derived class
   - Enables runtime polymorphism
   - Implemented through vtable mechanism

3. **Pure Virtual Functions**:
   - Declared with = 0
   - Makes class abstract (cannot instantiate)
   - Must be implemented by derived classes

4. **Virtual Destructor**:
   - ALWAYS make destructor virtual in base classes
   - Prevents memory leaks when deleting through base pointer
   - Critical for proper cleanup

5. **Object Slicing**:
   - Occurs when assigning derived object to base object
   - Use pointers/references to avoid slicing
   - Pointers/references enable polymorphism

6. **vtable & vptr**:
   - vtable: Array of function pointers per class
   - vptr: Pointer in each object to its vtable
   - Runtime overhead minimal, benefits huge

Best Practices:
- Always virtual destructor in base classes
- Use 'override' keyword for clarity
- Prefer pointers/references for polymorphic behavior
- Use pure virtual for interface-like classes
- Program to abstractions, not implementations

When to Use:
✅ Need different behaviors for related types
✅ Want to add new types without changing existing code
✅ Building extensible systems
✅ Implementing design patterns (Strategy, Factory, etc.)

Compile-time vs Runtime:
- Compile-time: Faster, decided at compile time
- Runtime: Flexible, decided at runtime, slight overhead
*/

File: oop-fundamentals/7.polymorphism/ts/polymorphism.ts
//
// =======================================================
// 1. WHAT IS POLYMORPHISM?
// =======================================================
//
// Polymorphism = "Many forms"
// It allows the same interface to be used for different underlying forms (data types).
//
// Key formula:
// Polymorphism = Same Interface + Different Behaviors
//
// Benefits:
// - Code reusability (write generic code)
// - Flexibility (add new types without changing existing code)
// - Loose coupling (depend on abstractions, not concrete types)
// -Extensibility (easily add new implementations)
//
// Types in TypeScript:
// 1. Compile-time Polymorphism (Static)
//    - Function Overloading (limited in TS)
//    - Method Overloading via union types
//
// 2. Runtime Polymorphism (Dynamic)
//    - Method Overriding
//    - Interface implementation

//
// =======================================================
// 2. COMPILE-TIME POLYMORPHISM: FUNCTION OVERLOADING
// =======================================================
//
// TypeScript supports function overloading through overload signatures
// However, implementation is single with type checking

class Calculator {
  // Overload signatures
  add(a: number, b: number): number;
  add(a: string, b: string): string;
  add(a: number, b: number, c: number): number;

  // Implementation signature (must handle all overloads)
  add(a: number | string, b: number | string, c?: number): number | string {
    if (typeof a === "number" && typeof b === "number") {
      if (c !== undefined) {
        console.log("📊 add(number, number, number)");
        return a + b + c;
      }
      console.log("📊 add(number, number)");
      return a + b;
    }
    console.log("📊 add(string, string)");
    return String(a) + String(b);
  }
}

//
// =======================================================
// 3. RUNTIME POLYMORPHISM: METHOD OVERRIDING
// =======================================================
//
// TypeScript implements runtime polymorphism through inheritance
// Child classes can override parent methods

abstract class Shape {
  protected color: string;

  constructor(color: string) {
    this.color = color;
  }

  // Abstract method - must be implemented by subclasses
  abstract getArea(): number;

  // Concrete method - can be overridden
  draw(): void {
    console.log(`Drawing a shape with color: ${this.color}`);
  }
}

class Circle extends Shape {
  private radius: number;

  constructor(color: string, radius: number) {
    super(color);
    this.radius = radius;
  }

  getArea(): number {
    return Math.PI * this.radius ** 2;
  }

  // Override parent method
  draw(): void {
    console.log(
      `⭕ Drawing Circle | Color: ${this.color} | Area: ${this.getArea().toFixed(2)}`,
    );
  }
}

class Rectangle extends Shape {
  private width: number;
  private height: number;

  constructor(color: string, width: number, height: number) {
    super(color);
    this.width = width;
    this.height = height;
  }

  getArea(): number {
    return this.width * this.height;
  }

  draw(): void {
    console.log(
      `▭ Drawing Rectangle | Color: ${this.color} | Area: ${this.getArea()}`,
    );
  }
}

//
// =======================================================
// 4. INTERFACE-BASED POLYMORPHISM
// =======================================================
//
// TypeScript's interfaces enable structural polymorphism
// Any object matching the interface shape can be used

interface Printable {
  print(): void;
}

interface Saveable {
  save(filename: string): void;
}

class Document implements Printable, Saveable {
  constructor(
    private title: string,
    private content: string,
  ) {}

  print(): void {
    console.log(`=== ${this.title} ===`);
    console.log(this.content);
  }

  save(filename: string): void {
    console.log(`💾 Saving '${this.title}' to ${filename}`);
  }
}

class Image implements Printable {
  constructor(
    private name: string,
    private format: string,
  ) {}

  print(): void {
    console.log(`🖼️  Printing image: ${this.name}.${this.format}`);
  }
}

//
// =======================================================
// 5. ABSTRACT CLASSES
// =======================================================
//
// Abstract classes provide both interface and partial implementation
// Cannot be instantiated directly

abstract class Animal {
  protected name: string;

  constructor(name: string) {
    this.name = name;
  }

  // Abstract method - must implement
  abstract makeSound(): void;

  // Concrete method - shared implementation
  sleep(): void {
    console.log(`${this.name} is sleeping 💤`);
  }

  eat(): void {
    console.log(`${this.name} is eating 🍖`);
  }
}

class Dog extends Animal {
  constructor(name: string) {
    super(name);
  }

  makeSound(): void {
    console.log(`${this.name} says: Woof! Woof! 🐕`);
  }
}

class Cat extends Animal {
  constructor(name: string) {
    super(name);
  }

  makeSound(): void {
    console.log(`${this.name} says: Meow! Meow! 🐱`);
  }
}

//
// =======================================================
// 6. REAL-WORLD EXAMPLE: PAYMENT SYSTEM
// =======================================================
//

interface PaymentMethod {
  processPayment(amount: number): boolean;
  getType(): string;
}

class CreditCardPayment implements PaymentMethod {
  constructor(
    private accountId: string,
    private cardNumber: string,
  ) {}

  processPayment(amount: number): boolean {
    console.log(`💳 Processing $${amount} via Credit Card`);
    console.log(`   Card: ${this.cardNumber.substring(0, 4)}****`);
    return true;
  }

  getType(): string {
    return "Credit Card";
  }
}

class PayPalPayment implements PaymentMethod {
  constructor(
    private accountId: string,
    private email: string,
  ) {}

  processPayment(amount: number): boolean {
    console.log(`💰 Processing $${amount} via PayPal`);
    console.log(`   Email: ${this.email}`);
    return true;
  }

  getType(): string {
    return "PayPal";
  }
}

class CryptoPayment implements PaymentMethod {
  constructor(
    private accountId: string,
    private walletAddress: string,
  ) {}

  processPayment(amount: number): boolean {
    console.log(`₿ Processing $${amount} via Cryptocurrency`);
    console.log(`   Wallet: ${this.walletAddress.substring(0, 8)}...`);
    return true;
  }

  getType(): string {
    return "Cryptocurrency";
  }
}

//
// =======================================================
// 7. POLYMORPHIC BEHAVIOR
// =======================================================
//

function processTransaction(method: PaymentMethod, amount: number): void {
  console.log("\n--- Processing Transaction ---");
  console.log(`Payment Type: ${method.getType()}`);
  method.processPayment(amount);
  console.log("✅ Transaction complete!");
}

//
// =======================================================
// 8. GENERIC POLYMORPHISM
// =======================================================
//
// TypeScript generics enable compile-time polymorphism

class Stack<T> {
  private items: T[] = [];

  push(item: T): void {
    this.items.push(item);
  }

  pop(): T | undefined {
    return this.items.pop();
  }

  peek(): T | undefined {
    return this.items[this.items.length - 1];
  }

  isEmpty(): boolean {
    return this.items.length === 0;
  }

  size(): number {
    return this.items.length;
  }
}

//
// =======================================================
// 9. DUCK TYPING (STRUCTURAL POLYMORPHISM)
// =======================================================
//
// TypeScript uses structural typing: if it looks like a duck, it's a duck

interface Flyable {
  fly(): void;
}

class Bird {
  fly(): void {
    console.log("Bird is flying! 🐦");
  }
}

class Airplane {
  fly(): void {
    console.log("Airplane is flying! ✈️");
  }
}

function makeFly(thing: Flyable): void {
  thing.fly(); // Works with ANY object that has fly() method!
}

//
// =======================================================
// 10. UNION TYPES FOR POLYMORPHISM
// =======================================================
//
// Union types allow variables to have multiple types

type Result = Success | Error;

class Success {
  constructor(public value: any) {}

  isSuccess(): boolean {
    return true;
  }
}

class Error {
  constructor(public message: string) {}

  isSuccess(): boolean {
    return false;
  }
}

function handleResult(result: Result): void {
  if (result.isSuccess()) {
    console.log(`✅ Success: ${(result as Success).value}`);
  } else {
    console.log(`❌ Error: ${(result as Error).message}`);
  }
}

//
// =======================================================
// 11. METHOD OVERRIDING WITH SUPER
// =======================================================
//

class Employee {
  constructor(
    protected name: string,
    protected salary: number,
  ) {}

  work(): void {
    console.log(`${this.name} is working...`);
  }

  getInfo(): string {
    return `${this.name} earns $${this.salary}`;
  }
}

class Manager extends Employee {
  constructor(
    name: string,
    salary: number,
    private teamSize: number,
  ) {
    super(name, salary);
  }

  work(): void {
    super.work(); // Call parent method
    console.log(`${this.name} is managing ${this.teamSize} people`);
  }

  getInfo(): string {
    return super.getInfo() + ` and manages ${this.teamSize} people`;
  }
}

//
// =======================================================
// 12. REAL-WORLD EXAMPLE: NOTIFICATION SYSTEM
// =======================================================
//

abstract class Notification {
  protected recipient: string;
  protected message: string;
  protected timestamp: Date;

  constructor(recipient: string, message: string) {
    this.recipient = recipient;
    this.message = message;
    this.timestamp = new Date();
  }

  formatHeader(): string {
    return `[${this.timestamp.toISOString()}] To: ${this.recipient}`;
  }

  abstract send(): void;
}

class EmailNotification extends Notification {
  constructor(
    recipient: string,
    message: string,
    private subject: string,
  ) {
    super(recipient, message);
  }

  send(): void {
    console.log("\n📧 EMAIL NOTIFICATION");
    console.log(this.formatHeader());
    console.log(`Subject: ${this.subject}`);
    console.log(`Body: ${this.message}`);
    console.log("✅ Email delivered");
  }
}

class SMSNotification extends Notification {
  private static readonly MAX_LENGTH = 160;

  constructor(
    recipient: string,
    message: string,
    private phoneNumber: string,
  ) {
    super(recipient, message);
  }

  send(): void {
    console.log("\n📱 SMS NOTIFICATION");
    console.log(this.formatHeader());
    console.log(`Phone: ${this.phoneNumber}`);

    const smsBody =
      this.message.length > SMSNotification.MAX_LENGTH
        ? this.message.substring(0, SMSNotification.MAX_LENGTH - 3) + "..."
        : this.message;

    console.log(`SMS: ${smsBody}`);
    console.log(
      `✅ SMS sent (${smsBody.length}/${SMSNotification.MAX_LENGTH} chars)`,
    );
  }
}

class PushNotification extends Notification {
  constructor(
    recipient: string,
    message: string,
    private deviceToken: string,
    private priority: string,
  ) {
    super(recipient, message);
  }

  send(): void {
    console.log("\n🔔 PUSH NOTIFICATION");
    console.log(this.formatHeader());
    console.log(`Device: ${this.deviceToken.substring(0, 8)}...`);
    console.log(`Priority: ${this.priority}`);
    console.log(`Alert: ${this.message}`);
    console.log("✅ Push delivered");
  }
}

//
// =======================================================
// 13. DEMONSTRATION
// =======================================================
//

function main(): void {
  console.log("=== Polymorphism in TypeScript Demo ===\n");

  // ---- Function Overloading ----
  console.log("1. Function Overloading (Compile-time):");
  const calc = new Calculator();
  console.log(`Result: ${calc.add(5, 3)}`);
  console.log(`Result: ${calc.add(5, 3, 2)}`);
  console.log(`Result: ${calc.add("Hello", " World")}`);

  // ---- Method Overriding ----
  console.log("\n2. Method Overriding (Runtime Polymorphism):");
  const circle = new Circle("Blue", 5);
  const rectangle = new Rectangle("Green", 10, 20);

  const shapes: Shape[] = [circle, rectangle];
  shapes.forEach((shape) => shape.draw());

  // ---- Interface-based Polymorphism ----
  console.log("\n3. Interface-based Polymorphism:");
  const doc = new Document("My Document", "This is important content.");
  const img = new Image("photo", "jpg");

  const printables: Printable[] = [doc, img];
  printables.forEach((p) => p.print());

  // ---- Abstract Classes ----
  console.log("\n4. Abstract Classes:");
  const dog = new Dog("Buddy");
  const cat = new Cat("Whiskers");

  const animals: Animal[] = [dog, cat];
  animals.forEach((animal) => {
    animal.makeSound();
    animal.sleep();
  });

  // ---- Payment System ----
  console.log("\n5. Real-World Example (Payment System):");
  const creditCard = new CreditCardPayment("ACC001", "1234567890123456");
  const paypal = new PayPalPayment("ACC002", "user@example.com");
  const crypto = new CryptoPayment(
    "ACC003",
    "0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb",
  );

  const payments: PaymentMethod[] = [creditCard, paypal, crypto];
  payments.forEach((payment) => processTransaction(payment, 100.0));

  // ---- Generic Polymorphism ----
  console.log("\n6. Generic Polymorphism (Stack):");
  const numberStack = new Stack<number>();
  numberStack.push(1);
  numberStack.push(2);
  numberStack.push(3);
  console.log(`Stack size: ${numberStack.size()}`);
  console.log(`Popped: ${numberStack.pop()}`);

  const stringStack = new Stack<string>();
  stringStack.push("Hello");
  stringStack.push("World");
  console.log(`Stack size: ${stringStack.size()}`);
  console.log(`Popped: ${stringStack.pop()}`);

  // ---- Duck Typing ----
  console.log("\n7. Duck Typing (Structural Polymorphism):");
  const bird = new Bird();
  const plane = new Airplane();

  makeFly(bird); // Works with Bird
  makeFly(plane); // Works with Airplane (structural typing!)

  // ---- Union Types ----
  console.log("\n8. Union Types:");
  const success = new Success(42);
  const error = new Error("Something went wrong");

  handleResult(success);
  handleResult(error);

  // ---- Method Overriding with Super ----
  console.log("\n9. Method Overriding with Super:");
  const manager = new Manager("Alice", 120000, 5);
  manager.work();
  console.log(manager.getInfo());

  // ---- Notification System ----
  console.log("\n10. Real-World Example (Notification System):");
  const email = new EmailNotification(
    "alice@example.com",
    "Your order has been shipped!",
    "Order Update",
  );
  const sms = new SMSNotification(
    "Bob",
    "Your verification code is 482910.",
    "+1-555-0123",
  );
  const push = new PushNotification(
    "Charlie",
    "New message from Alice",
    "d8a3f4b2c1e5a9b7",
    "high",
  );

  const notifications: Notification[] = [email, sms, push];
  notifications.forEach((notification) => notification.send());

  console.log("\n=== Demo Complete ===");
}

// Execute
main();

export {};

/*
📘 Key Insights:

1. **TypeScript Polymorphism**:
   - Compile-time: Function overloading (limited), generics
   - Runtime: Method overriding, interface implementation
   - Structural typing: Duck typing (if it quacks like a duck...)

2. **Function Overloading**:
   - Overload signatures + single implementation
   - Less flexible than C++ but type-safe
   - Runtime checks needed in implementation

3. **Method Overriding**:
   - Child classes override parent methods
   - No special keyword needed (unlike C++'s virtual)
   - Runtime polymorphic behavior

4. **Interfaces**:
   - Define contracts
   - Structural typing (shape-based)
   - Multiple interface implementation
   - No runtime representation

5. **Abstract Classes**:
   - Cannot instantiate
   - abstract methods must be implemented
   - Can have concrete methods
   - Use when sharing implementation

6. **Generics**:
   - Compile-time polymorphism
   - Type-safe reusable code
   - Works with any type

7. **Duck Typing**:
   - Structural typing unique to TypeScript
   - If it has the right shape, it works
   - More flexible than nominal typing

Best Practices:
- Use interfaces for contracts
- Use abstract classes for shared behavior
- Leverage generics for type-safe code
- Use method overriding for runtime polymorphism
- Prefer composition over inheritance when possible

When to Use:
✅ Need different behaviors for related types
✅ Want to add new types without changing existing code
✅ Building extensible systems
✅ Type-safe generic code needed

TypeScript Advantages:
- Type safety at compile time
- Structural typing flexibility
- Generics for reusable code
- Clean syntax (no virtual keyword needed)
*/
